core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Override
+    public boolean isOpen() {
+        return source.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return source.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return source.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return source.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        source.subscribe(new AsyncMapSubscriber<>(subscriber, function), executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class AsyncMapSubscriber<U> implements Subscriber<Object>, Subscription {
+        private final Subscriber<? super U> downstream;
+        private final Function<Object, CompletableFuture<U>> function;
+
+        @Nullable
+        private volatile Subscription upstream;
+        private volatile boolean canceled;
+
+        AsyncMapSubscriber(Subscriber<? super U> downstream,
+                           Function<Object, CompletableFuture<U>> function) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(function, "function");
+            this.downstream = downstream;
+            this.function = function;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(Object item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            try {
+                final CompletableFuture<U> future = function.apply(item);
+                requireNonNull(future, "function.apply() returned null");
+                future.whenComplete((res, cause) -> {
+                    try {
+                        if (cause != null) {
+                            upstream.cancel();
+                            downstream.onError(cause);
+                        } else {
+                            requireNonNull(res, "function.apply()'s future completed with null");
+                            downstream.onNext(res);
+                        }
+                    } catch (Throwable ex) {
+                        StreamMessageUtil.closeOrAbort(item, ex);
+                        upstream.cancel();
+                        onError(ex);
+                    }
```

Comment:
- We need to publish the `res` using the `executor` specified when subscribing to this stream.
- There is a performance problem in `CompletableFuture` of old JDK versions. We prefer `handle` for handling a result and an error in a callback.
- We must not send an error more than once to down stream. If a stream was canceled or a error was raised before completing the transforming function,
  additionally operation should be stopped. 
```java
future.handle((res, cause) -> {
    if (executor.inEventLoop()) {
        onNext0(res, cause);
    } else {
       executor.execute(() -> onNext(res, cause);
    }
    return null;
});

void onNext0(@Nullable U item, @Nullable Throwable cause) {
    if (canceled) {
        // A subscription was canceled.
        return;
    }
    ...
}
```

Reply:
I think I have this all correctly now 🙂